hugr/src/hugr/validate.rs

```diff
@@ -51,15 +48,52 @@ impl Hugr {
         self.validate_with_extension_closure(HashMap::new(), extension_registry)
     }
 
+    /// Check the validity of the HUGR, but don't check consistency of extension
+    /// requirements between connected nodes or between parents and children.
+    pub fn validate_no_extensions(
+        &self,
+        extension_registry: &ExtensionRegistry,
+    ) -> Result<(), ValidationError> {
+        let mut validator = ValidationContext::new(self, extension_registry);
+        validator.validate()
+    }
+
+    pub fn validate_extensions(&self, closure: ExtensionSolution) -> Result<(), ValidationError> {
+        let validator = ExtensionValidator::new(self, closure);
+        for src_node in self.nodes() {
+            let node_type = self.get_nodetype(src_node);
+
+            // FuncDefns have no resources since they're static nodes, but the
+            // functions they define can have any extension delta.
+            if node_type.tag() != OpTag::FuncDefn {
+                // If this is a container with I/O nodes, check that the extension they
+                // define match the extensions of the container.
+                if let Some([input, output]) = self.get_io(src_node) {
+                    validator.validate_io_extensions(src_node, input, output)?;
+                }
+            }
+
+            for src_port in self.node_outputs(src_node) {
+                for (tgt_node, tgt_port) in self.linked_inputs(src_node, src_port) {
+                    validator.check_extensions_compatible(&(src_node, src_port.into()), &(tgt_node, tgt_port.into()))?;
+                }
+            }
+        }
+        Ok(())
+    }
+
     /// Check the validity of a hugr, taking an argument of a closure for the
     /// free extension variables
     pub fn validate_with_extension_closure(
         &self,
         closure: ExtensionSolution,
         extension_registry: &ExtensionRegistry,
     ) -> Result<(), ValidationError> {
-        let mut validator = ValidationContext::new(self, closure, extension_registry);
-        validator.validate()
+        let mut validator = ValidationContext::new(self, extension_registry);
+        validator.validate()?;
+        #[cfg(feature = "extension_inference")]
```

Comment:
Is it plausible to move this #[cfg] onto the `fn validate_with_extension_closure` itself?

Reply:
Hmmmm, *or*, leave this as a way of validating extensions *even if `extension_inference` is disabled*? So then the feature-flag doesn't disable any validation, but just changes what the default `validate` and `update_validate` do?

Reply:
I've left this as is, updating `Hugr::validate` to point to different things depending on the flag status instead